pkg/project/project.go

```diff
@@ -162,3 +162,147 @@ func Exists(client *occlient.Client, projectName string) (bool, error) {
 	}
 	return false, errors.Errorf(" %v project does not exist", projectName)
 }
+
+type Project struct {
+	Name   string
+	Active bool
+	Client *occlient.Client
+}
+
+func (p *Project) Create() error {
+	err := p.Client.CreateNewProject(p.Name)
+	if err != nil {
+		return errors.Wrap(err, "unable to create new project")
+	}
+	return nil
+}
+
+func (p *Project) Exists() (bool, error) {
+	project, err := p.Client.GetProject(p.Name)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to get the project")
+	}
+	if project == nil {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (p *Project) SetActive() error {
+	currentProject := GetCurrent(p.Client)
+
+	cfg, err := config.New()
+	if err != nil {
+		return errors.Wrap(err, "unable to access config file")
+	}
+	err = cfg.UnsetActiveApplication(currentProject)
+	if err != nil {
+		return errors.Wrap(err, "unable to unset active application of current project "+p.Name)
+	}
+	err = cfg.UnsetActiveComponent(currentProject)
+	if err != nil {
+		return errors.Wrap(err, "unable to unset active component of current project "+p.Name)
+	}
+
+	err = p.Client.SetCurrentProject(p.Name)
+	if err != nil {
+		return errors.Wrap(err, "unable to set current project to"+p.Name)
+	}
+	return nil
+}
+
+func (p *Project) Delete() (string, error) {
+	// Loading spinner
+	s := log.Spinnerf("Deleting project %s", p.Name)
+	defer s.End(false)
+
+	currentProject := GetCurrent(p.Client)
+
+	projects, err := List(p.Client)
+	if err != nil {
+		return "", errors.Wrapf(err, "unable to fetch list of projects")
+	}
+
+	//Iterate the project list and see the expected change post deletion
+	for ind, prj := range projects {
+		if prj.Name == p.Name {
+			projects = append(projects[:ind], projects[ind+1:]...)
+		}
+	}
+
+	// If current project is not same as the project to be deleted, set it as current
+	if currentProject != p.Name {
+		// Set the project to be deleted as current inorder to be able to delete it
+		err = SetCurrent(p.Client, p.Name)
+		if err != nil {
+			return "", errors.Wrapf(err, "Unable to delete project %s", p.Name)
+		}
+	}
+
+	// Delete the requested project
+	err = p.Client.DeleteProject(p.Name)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to delete project")
+	}
+
+	// delete from config
+	cfg, err := config.New()
+	if err != nil {
+		return "", errors.Wrapf(err, "unable to delete project from config file")
+	}
+
+	err = cfg.DeleteProject(p.Name)
```

Comment:
If I am not wrong shouldn't we delete the project first from the cfg and then create a new config?

Reply:
Ok 2 things here:

1. Update of config file is a local file(created by odo at the moment) update
    a. No n/w access
    b. possible failure cause: disk failure
2. Deletion of project is a request to the cluster
    a. N/w failure cutting off access to the (possibly remote) cluster
    b. incorrect parameters to the request -- Unlikely, considering they need to be pre-validated by odo before proceeding to here...

Considering the above, if we update the config first and then the delete request to the cluster, I feel it would lead us to inconsistent behaviour... Whereas, if you make the delete request and then updation to the config file fails due to say a local disk failure, there is hardly any recovery mechanism until the disk failure is fixed so in essence, this inconsistency can hardly be met AFAIK but the other common case above is dealt with

So I tend to favor the approach of deleting the project first and then trying an update on config file